### DISCUSSION TITLE
Migrate away form cocoapods on iOS

### DISCUSSION THREAD

**Discussion by lemcoder** [GitLiveApp/firebase-kotlin-sdk#573] [Ideas]
Right now to integrate with iOS one must use a cocoapods plugin. This is quite heavy dependency especially in case of build time.
I would like to propose a more lightweight solution that is packing the static framework within a library and then use a cinterop plugin to generate kotlin bindings. This way the requirement to use cocoapods would be lifted.

Let me know if it is something that You would consider a good approach and I shall start working on the PR 

Thanks! 🙌

**Comment by nbransby** [comment]
thoughts? @Reedyuk @Daeda88

**Reply by Reedyuk** [reply]
Feels similar to how we had before with Carthage.
I would stick to using the tools that jetbrains have given us, rather than roll our own solution.

**Reply by Daeda88** [reply]
Yeah we moved away from Carthage because we were seeing issues with building it now that larger parts of the firebase sdk are Swift first. I agree Cocoapods has a lot of downsides in terms of build speed but it's difficult to resolve.

What we at Splendo have done to reduce build times dramatically is to make a cache of the Cocoapods and some gradle plugin to import it. In theory we could make a similar plugin for this library.

**Comment by Qw4z1** [comment]
@lemcoder I'm a novice in this area, but are you talking about having to use cocoapods to build the library itself and using something else to integrate Firebase iOS into the library? Of so, I'm at least interested in looking at an example of what this could look like. 😊

**Reply by lemcoder** [reply]
@Qw4z1 I was thinking about using the cinterop tool whitch is the `tool that jetbrains have given us`.
More context: https://kotlinlang.org/docs/native-c-interop.html#create-bindings-for-a-new-library
You can find some talk here: https://www.youtube.com/watch?v=w_mkLrzb_I4&t=2197s

Also I did something similar lately but on Windows with the https://github.com/schellingb/TinySoundFont library
Basically it boils down to obtaining (either compiling from source or downloading i.e from cocoapods) the static library. Then you configure the cinterop to generate kotlin bindings based on some header file (`.h`). Lastly you point the cinterop tool to bundle the static lib with your app/package.

@Daeda88 If I am not mistaken the cocoapods plugin largely relies on cinterop plugin so the SDKs beeing written in Swift are as much of a problem when using cocoapods.

I haven't dived into code yet but it seems that `firebase-kotlin-sdk` is split into modules that are separate libraries per firebase module. That would actually simplify things because we would need to add only one static library per module and the download size would not increase that much.

The only thing that I am not sure of is how to obtain the static versions of firebase iOS frameworks. But I belive they are somwhere inside the `pod`

**Comment by Daeda88** [comment]
To be clear, whatever approach we take here, it does not rule out anyone using a different one in their project. If we use a different setup it would improve build speed of the github actions a bit, but it wont help anyone implementing this sdk as these dependencies are not transitive. Cocoapods simply sets up all the library linking, including any sub-dependencies. With Carthage this was a manual process and that proved trickier than just using the cocoapods plugin instead.

**Reply by Qw4z1** [reply]
Yes, that's what I assumed. I'm using cocoapods for one of my projects and SPM for  another, and would hate to have to switch to a third. 😅

**Reply by lemcoder** [reply]
The dependencies could be transitive if linked statically with the output `.klib`
I can provide an example to demonstrate how this works. Actually I have example at hand but it is written for windows so let me know if it is not a problem

**Reply by lemcoder** [reply]
[pandamidi-master.zip](https://github.com/user-attachments/files/16380324/pandamidi-master.zip)

Here is the example for mingwx64 target. To run please build the `:core:` package first. Then run `publishAllPublicationsToMavenLocalRepository`. Finally change the path to a file in the main function (you can get the file here https://github.com/schellingb/TinySoundFont/blob/main/examples/florestan-subset.sf2) Then run the main function from the gutter icon

The output should be `Presets: 17` 

As you can see the app is relying only on the kotlin library. No need to link it with the `libtsf.a` because it was linked statically. 
https://blog.hubspot.com/website/static-vs-dynamic-linking

**Comment by nbransby** [comment]
> Right now to integrate with iOS one must use a cocoapods plugin.

To clarify,  do you mean as a contributor to this library (as in you have cloned this repo and building it locally) or as a consumer of this library (as it you include the dev.gitlive.* maven artifacts in your own project)?

**Reply by lemcoder** [reply]
I mean as consumer of library I have to provide the implementation of iOS Firebase SDK myself either via cocoa pods or SPM and I would love to just import the library in my common code and for it to work on iOS "batteries included". 

You can see how Ktor is doing It with native (curl) HTTP clients [here](https://github.com/ktorio/ktor/tree/main/ktor-client/ktor-client-curl/desktop). And when you use it you do not need to link curl with your app

**Reply by nbransby** [reply]
That would certainly provide a much more frictionless setup but is it possible for iOS libraries? Also are you still able to access the iOS sdk directly from kotlin iOS actuals and swift code in the case you need to?

**Reply by lemcoder** [reply]
I will do my best to provide a small PR for one of the modules in the near future. It will help in testing and evaluating the idea. Linking the iOS libraries [is possible](https://kotlinlang.org/docs/multiplatform-ios-dependencies.html#without-cocoapods), as well as accessing the generated Kotlin code from iOS actuals.

Regarding accessing Swift code, I do not think it will be possible, as the implementations will be shipped within the `firebase-kotlin-sdk` library. Thus, consumers will have access only to the parts exposed by the library.
